File: 020-ValidParentheses/c++/main.cpp
#include <vector>
#include <string>
#include <sstream>
#include <iostream>
#include "common/cpp_common.h"
#include <algorithm>
#include <map>
#include <climits>
#include <unordered_map>
#include <set>
#include <stack>
using namespace std;

// https://leetcode.com/problems/valid-parentheses/
// 这个问题使用栈来解决，最后判断栈是不是为空
class Solution {
public:
    Solution() {
        mymap[')'] = '(';
        mymap[']'] = '[';
        mymap['}'] = '{';
        myset.insert('(');
        myset.insert('[');
        myset.insert('{');
    }
    // 解法一， 使用了额外的 map 和 set 空间
    bool isValid(string s) {
        std::stack<char> mystack;
        for (int i = 0; i < s.length(); i++) {
            if (myset.find(s[i]) != myset.end()) mystack.push(s[i]);
            else {
                if (mystack.empty()) return false;
                char top = mystack.top();
                if (mymap[s[i]] != top) 
                    return false;
                else
                    mystack.pop();
            }
        }
        if (mystack.empty()) return true;
        else return false;
    }
    // 解法二： switch case 节省空间
    bool isValid_2(string s) {
        std::stack<char> mystack;
        for (int i = 0; i < s.length(); i++) {
            switch(s[i]) {
                case '(': mystack.push(']'); break;
                case '{': mystack.push('}'); break;
                case '[': mystack.push(']'); break;
                default:
                    if (mystack.empty() || mystack.top() != s[i]) return false;
                    else mystack.pop();
            }
        }
        return mystack.empty();
    }
private:
    std::map<char, char> mymap;
    std::set<char> myset;
};

int main() {
    std::vector<std::string> strs;
    std::string str;
    leetcode::common::get_string_from_file("./data1.txt", str);
    strs = leetcode::common::split(str,'\n');
    Solution *solution = new Solution(); 
    for (auto str : strs) {
        cout << str << endl;
        auto result = solution->isValid_2(str);
        cout << (result ? "True" : "False") << endl;
    }
   
    return 0;
}

File: 020-ValidParentheses/go/main.go
package main

import (
	"common"
	"fmt"
)

func twoSum(nums []int, target int) []int {
	var ret []int
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				ret = append(ret, i)
				ret = append(ret, j)
				return ret
			}
		}
	}
	return ret
}

func twoSum_2(nums []int, target int) []int {
	var ret []int
	var hmap map[int]int = make(map[int]int, len(nums))
	for i := 0; i < len(nums); i++ {
		hmap[nums[i]] = i
	}
	for i := 0; i < len(nums)-1; i++ {
		sub := target - nums[i]
		index, ok := hmap[sub]
		if ok && index != i {
			ret = append(ret, i)
			ret = append(ret, index)
			return ret
		}
	}
	return ret
}

func main() {
	str := common.GetStringFromFile("./data1.txt")
	strs := common.Split(str, "\n")

	nums, err := common.StringToInt(strs)
	if err != nil {
		fmt.Println(err.Error())
		return
	}
	ret := twoSum(nums, 9)
	fmt.Println(ret)
	ret = twoSum_2(nums, 9)
	fmt.Println(ret)
}
